### PR TITLE
Follow up execution normalization, SL protection, and watermark boundaries

### PR DESCRIPTION
**Pull request by wuyaocheng**
## Summary
- add execution normalization telemetry so raw sizing vs normalized exchange order parameters are observable
- make SL protection pricing depth-aware with top-of-book coverage telemetry
- separate watermark refresh from live position/exit state derivation for cleaner trailing-stop boundaries

## Testing
- go test ./internal/service/...
- go test ./...

### CODE REVIEW

**Comment by wuyaocheng** [comment]
@codex review

File: internal/service/execution_strategy.go [hunk 1]
```diff
@@ -571,31 +590,72 @@ func resolvePassiveBookPrice(side string, bestBid, bestAsk, fallback float64) fl
...
+				spreadCapped := math.Min(bestAsk, bestBid+allowedCross)
+				if quantity > 0 && bestAskQty > 0 {
+					coverage := math.Min(bestAskQty/quantity, 1)
+					decision.ExpectedCoverage = coverage
+					if coverage >= 1 {
+						decision.Price = bestAsk
+						decision.DepthMode = "top-book-cover"
+						return decision
+					}
+					decision.Price = spreadCapped + (bestAsk-spreadCapped)*(1-coverage)
```
[the remark this answers is not in the document]
Comment (folgercn) [reply internal/service/execution_strategy.go]: 这次 follow-up 整体方向是对的，我这边看下来倾向通过。

主要加分点有几点：

execution normalization telemetry 补得比较完整。现在从理论 quantity / price 到交易所实际提交参数之间的变化链路更清楚了，后续排查是 sizing 决策问题还是交易所规则裁剪问题，会方便很多。
SL protection 这次增加了基于 top-of-book 覆盖率的 telemetry，虽然目前还不是完整 depth-aware，但已经比单纯 spread cap 更可观测，后续继续往多档 depth 扩展也有基础。
trailing stop / watermark 相关逻辑这次把状态刷新和状态推导拆得更清楚了。这样对 live recovery、单元测试、后续维护都更友好。

有两个非阻塞建议，后续可以继续优化：

normalization 这块虽然 telemetry 已经补齐，但如果后面要进一步做精细复盘，建议把 step size 调整前、min notional 二次放大后的数量链路再表达得更显式一点。
coverage-weighted-cap 目前本质还是 heuristic，建议后续在注释或文档里再明确一下它是保护限价估算，不代表真实成交均价模型。

总体来说，这个 PR 在执行链路可观测性和 live 状态边界上都有明显提升，可以合并。

**Comment by wuyaocheng** [comment]
Addressed the remaining BUY-side SL protection follow-up in `cf86436`:

- corrected the BUY branch interpolation in `resolveAggressiveSLProtectionDecision` so partial top-of-book coverage moves monotonically from `spreadCapped` toward `bestAsk` as coverage increases, instead of reversing the slippage cap
- added a symmetric BUY-side regression test for the coverage-weighted SL protection path

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Review by folgercn — APPROVED** [review]
补看了最新提交，这次修正是到点上的。

新增提交把 buy-side 的 `coverage-weighted-cap` 插值方向改对了，同时补了对应单测，和 sell-side 的语义现在一致了。我这里没有新的阻塞问题了。

整体看，这个 PR 在几个方面都比较扎实：
- execution normalization telemetry 更完整，理论下单参数与交易所归一化后的差异更容易排查；
- SL protection 增加了 top-of-book 覆盖相关 telemetry，可观测性更好；
- watermark 刷新与 live position / exit state 推导边界更清楚，测试和 recovery 更容易维护。

后续还能继续优化的点，还是之前那两个非阻塞建议：
- normalization 过程链路如果后面还要继续做精细复盘，可以再拆得更细一点；
- `coverage-weighted-cap` 目前本质还是 heuristic，后续可以在注释或文档里进一步强调边界。

这版我可以通过。

**Comment by wuyaocheng** [comment]
Addressed the latest AI follow-ups in `d3a9581`:

- tightened aggressive SL protection so the final LIMIT price never breaches the configured `spreadCapped` bound on either SELL or BUY paths, while still preserving depth telemetry
- merged timeout/sync `adapterSubmission` metadata field-by-field so partial cancellation payloads inherit missing normalization data from the original submission
- expanded the watermark position key to include more position context (`symbol`, `entryPrice`, and `updatedAt`/`quantity` when available) so a same-price reopen does not inherit stale HWM/LWM
- added regression coverage for BUY-side capped SL protection, partial adapterSubmission fallback, and the expanded watermark-key reset behavior

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Comment by wuyaocheng** [comment]
Addressed the newest AI review edge cases in `3f5d825`:

- made the aggressive SL protection modes explicit: when top-of-book is outside the slippage cap we now stay at `spreadCapped` (`top-book-outside-cap`) instead of pretending to interpolate toward an out-of-cap quote; when top-of-book is actually within cap we still use `top-book-cover`
- changed `adapterSubmission` fallback merging to preserve original normalization fields unless the newer payload provides a meaningful replacement, so partial timeout/cancel payloads no longer wipe valid sizing/price telemetry
- tightened watermark identity to prefer a stable position `id`, falling back to `symbol|side|entryPrice`, and removed `updatedAt` from the key so timestamp refreshes do not reset trailing protection
- updated regression coverage around capped SL paths, partial adapterSubmission merge behavior, and stable watermark-key resets

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Comment by folgercn** [comment]
补看了最新一轮 AI review 后，我这里改一下判断：这个 PR 目前我不建议直接合并。

我认为至少还有两个阻塞点需要先处理：

1. `resolveAggressiveSLProtectionDecision()` 当前 `coverage-weighted-cap` 的结果可能突破 `spreadCapped`。如果最终保护价还能落到 `maxSlippageBps` 允许范围之外，那么这条 SL protection 实际上就失去了“滑点封顶”的意义。这里我建议先把约束收紧，至少保证最终价格不会越过 `spreadCapped`。

2. `watermarkPositionKey` 现在只由 `side|entryPrice` 组成，无法区分“同方向、同入场价”的新旧仓位。实盘里平仓后在同价再次开仓是可能发生的，这里会复用上一笔仓位遗留的 `hwm/lwm`，对 trailing stop / protection 会有真实风险。建议把能唯一标识本次持仓的字段纳入 key。

另外，`adapterSubmission` fallback 目前只在整体为空时才回填，这个我觉得也值得补，但优先级低于前面两个。

这几个点修掉后我再看。

**Comment by wuyaocheng** [comment]
Followed up the newest PR #22 review signal in `d367a89`:

- adjusted `adapterSubmission` fallback merging so explicit scalar values like `0` / `false` are preserved instead of being treated as missing
- extended `TestNormalizeRESTOrderRecordsNormalizationTelemetry` to assert the normalized order's actual `Quantity` and `Price`, not just the telemetry payload

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Comment by wuyaocheng** [comment]
Pushed one more follow-up in `df355ad` to tighten watermark identity on the non-real-position side too:

- virtual live positions now record a stable synthetic `id` when they are created
- watermark resolution already prefers a position `id`, so virtual/recovered virtual paths now avoid falling back to the weaker `symbol|side|entryPrice` identity
- added regression coverage to ensure virtual snapshots preserve that id and that watermark resets follow the new virtual identity

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Comment by wuyaocheng** [comment]
Addressed the latest execution follow-ups in `848334e`:

- stabilized `adapterSubmission` fallback semantics so explicit `false` and explicit `0` values are preserved by default, while still treating zeroed `symbolRules.*` snapshots as missing fallback fields
- tightened watermark identity/lifecycle so active positions prefer stable IDs, inactive positions clear persisted watermarks, and legacy virtual keys are only used for compatibility reads before upgrading to the canonical key
- added regression coverage around explicit-zero fallback semantics, legacy/virtual watermark compatibility, inactive-position clearing, and stable same-position watermark reuse

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Comment by wuyaocheng** [comment]
@codex review

**Comment by folgercn** [comment]
这轮跟进我看过了，整体没问题。

`adapterSubmission` 的 fallback 语义现在更一致了：显式传入的 `false` 和 `0` 都能按预期保留，不会再被旧值覆盖；同时对 `symbolRules.*` 这类零值快照字段，仍然保留了缺失时的回退行为。

watermark 这块也比之前更稳了：活跃仓位优先使用稳定的 canonical key，非活跃仓位会清理持久化 watermark 状态，legacy virtual key 也只用于兼容读取，之后会升级到新的标准 key。

回归测试覆盖也补得比较到位，这次新增的几个边界场景基本都照顾到了。

我这边没有新的阻塞问题，可以继续。

**Comment by wuyaocheng** [comment]
Addressed the latest watermark/fallback follow-ups in `3cc8a68`:

- restored execution fallback semantics so zeroed normalization fields (`rawQuantity`, `normalizedQuantity`, `rawPriceReference`, `normalizedPrice`, including nested normalization fields) are treated as missing and fall back to the original submission, while explicit `false` booleans still remain explicit
- tightened watermark lifecycle in `evaluateLiveSignalDecision(...)` so inactive snapshots clear persisted `hwm/lwm/watermarkPositionKey` immediately instead of leaking the previous position's trailing state into the next entry
- restored `entryPrice` to the non-ID watermark base key to avoid same-symbol/same-side reopen collisions, while preserving legacy real-position (`SIDE|ENTRY`) and legacy virtual keys during migration
- added regression coverage for BUY within-cap partial depth, legacy real-position key compatibility, and the revised normalization-fallback semantics

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Comment by wuyaocheng** [comment]
@codex review

**Comment by wuyaocheng** [comment]
Addressed the latest two follow-ups in `e732c9b`:

- fallback booleans now preserve explicit `false` only for unknown fields, while known execution-control flags (`postOnly`, `reduceOnly`, `closePosition`, `slProtectionActive`) still fall back to the original submission when timeout/cancel payloads only contain default zero values
- inactive snapshots no longer clear persisted watermarks in the generic refresh path, so transient missing-position reads do not silently loosen existing trailing/protection state

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Comment by wuyaocheng** [comment]
@codex review

**Comment by wuyaocheng** [comment]
@codex review

**Comment by wuyaocheng** [comment]
Addressed the latest fallback/watermark follow-ups in `bc02d7a`:

- preserved explicit `false` booleans during adapterSubmission fallback while still treating zeroed quantity fields as missing and keeping zeroed price-reference fields explicit
- cleared persisted watermark state in `EvaluateSignal(...)` when the current snapshot is inactive, without regressing the generic refresh path for transient missing-position reads
- tightened watermark key migration tests around canonical real/virtual keys and inactive-position clearing

Validation:
- `go test ./internal/service/...`
- `go test ./...`

**Comment by folgercn** [comment]
我重新看了最新这次提交，整体比上一轮稳定很多，之前我担心的两个点这次已经修回来了：

1. 布尔字段的 fallback 语义重新对齐了，显式传入的 `false` 现在会按“已提供值”处理，不会再被旧值覆盖；这点和测试里的 `reduceOnly=false` 预期也已经一致。
2. inactive position 时的 watermark 清理逻辑也恢复了，而且这次补了 strategy evaluation 层面的测试，能覆盖掉上一笔仓位状态残留到后续评估的问题。

另外，legacy watermark key 向 canonical key 的迁移意图这版也更清楚了，整体契约比上一版稳定。

唯一还想确认一下的是：`normalizedPrice` / `rawPriceReference` 现在的零值语义已经改成“显式 0 也保留，不再 fallback”。这个改动不一定有问题，但它相当于调整了 execution submission contract。只要这个是有意设计，我这边没有其他阻塞问题。
